packages/miniflare/src/plugins/kv/assets.ts

```diff
@@ -97,3 +98,122 @@ export function getAssetsServices(options: AssetsOptions): Service[] {
 	};
 	return [storageService, namespaceService];
 }
+
+const MAX_ASSET_COUNT = 20_000;
+const MAX_ASSET_SIZE = 25 * 1024 * 1024;
+const MANIFEST_HEADER_SIZE = 20;
+
+const PATH_HASH_OFFSET = 0;
+const PATH_HASH_SIZE = 16;
+
+const CONTENT_HASH_OFFSET = PATH_HASH_SIZE;
+const CONTENT_HASH_SIZE = 16;
+
+const TAIL_RESERVED_SIZE = 8;
+
+const ENTRY_SIZE = PATH_HASH_SIZE + CONTENT_HASH_SIZE + TAIL_RESERVED_SIZE;
+
+type AssetManifestEntry = { contentHash: Uint8Array; pathHash: Uint8Array };
+
+const walk = async (dir: string) => {
+	const files = await fs.readdir(dir, { recursive: true });
+	const manifest: AssetManifestEntry[] = [];
+	let counter = 0;
+	await Promise.all(
+		files.map(async (file) => {
+			const filepath = path.join(dir, file);
+			const relativeFilepath = path.relative(dir, filepath);
+			const filestat = await fs.stat(filepath);
+
+			if (filestat.isSymbolicLink() || filestat.isDirectory()) {
+				return;
+			} else {
+				if (counter >= MAX_ASSET_COUNT) {
+					throw new Error(
+						`Maximum number of assets exceeded.\n` +
+							`Cloudflare Workers supports up to ${MAX_ASSET_COUNT.toLocaleString()} assets in a version. We found ${counter.toLocaleString()} files in the specified assets directory "${dir}".\n` +
+							`Ensure your assets directory contains a maximum of ${MAX_ASSET_COUNT.toLocaleString()} files, and that you have specified your assets directory correctly.`
+					);
+				}
+
+				if (filestat.size > MAX_ASSET_SIZE) {
+					throw new Error(
+						`Asset too large.\n` +
+							`Cloudflare Workers supports assets with sizes of up to ${MAX_ASSET_SIZE}. We found a file ${filepath} with a size of ${filestat.size}.\n` +
+							`Ensure all assets in your assets directory "${dir}" conform with the Workers maximum size requirement.`
+					);
+				}
+
+				// not the same as deploy but more similar to what is stored in gcs
+				manifest.push({
+					contentHash: fakeHashFile(),
```

Comment:
Perhaps just this?

```suggestion
					contentHash: relativeFilepath,
```

Reply:
see https://github.com/cloudflare/workers-sdk/pull/6525#discussion_r1722039312